### PR TITLE
Describe the file demo in the README

### PR DESCRIPTION
**Pull request by OscarShiang**
Create README file in demos/file to explain concepts of the file demo.

### CODE REVIEW

File: docs/demos.md [hunk 1]
```diff
@@ -31,4 +31,4 @@ For the best understanding we suggest that you:
 | trace                                                       | Redirecting another thread by writing to its PC.                                  |
 | signalhandling                                              | Installing and invoking signal handlers.                                          |
 | backtrace                                                   | Show callstack of user threads. (only tested on Arm -O0)                          |
-| fiber                                                       | User space threading.                                                             |
\ No newline at end of file
+| fiber                                                       | User space threading.                                                             |
```
Comment (DavidSpickett) [inline docs/demos.md]: Looks like an accidental edit, not that I can tell from the diff what changed here.

File: docs/demos.md [hunk 2]
```diff
@@ -21,7 +21,7 @@ For the best understanding we suggest that you:
 | threadlocalstorage (Arm/Thumb)                              | Using thread local storage (TLS) to give each thread it's own 'global' variables. |
 | [conditionvariables](../demos/conditionvariables/README.md) | Waiting on, signalling and broadcasting to condition variables.                   |
 | [cancel](../demos/cancel/README.md)                         | Cancelling threads.                                                               |
-| file                                                        | Read from a file.                                                                 |
+| [file](../demos/file/READM.md)                              | Read from a file.                                                                 |
```
Comment (DavidSpickett) [inline docs/demos.md]: README.md (missing the last E)

File: demos/file/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+* `close` (in `src/user/file.c`)
+* `read` (in `src/user/file.c`)
+* `write` (in `src/user/file.c`)
+* `remove` (in `src/user/file.c`)
+
+## `open`
+
+This function opens a file descriptor according to the `path` and the `oflags` we specified.
+
+In AMT, the permissions we provide are `O_RDONLY` and `O_WRONLY`.
```
Comment (DavidSpickett) [inline demos/file/README.md]: Although these are kinda permissions, https://man7.org/linux/man-pages/man2/open.2.html refers to them as "access modes" so I'd rather use that term.

File: demos/file/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+If we can open a file, the return value will be nonzero. If we can not open a file, the value will be set to `-1`.
+
+## `close`
+
+We can close a file descriptor by using this function. 
+
+If the operation succeeds, the return value will be `0`. If the operation failed, the return value will be `-1`
```
Comment (DavidSpickett) [inline demos/file/README.md]: you can remove the first comma

"If the operation succeeds the return value will be `0`."

File: demos/file/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+* `write` (in `src/user/file.c`)
+* `remove` (in `src/user/file.c`)
+
+## `open`
+
+This function opens a file descriptor according to the `path` and the `oflags` we specified.
+
+In AMT, the permissions we provide are `O_RDONLY` and `O_WRONLY`.
+
+If we can open a file, the return value will be nonzero. If we can not open a file, the value will be set to `-1`.
```
Comment (DavidSpickett) [inline demos/file/README.md]: "open the file" makes more sense here

File: demos/file/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+We can use `write()` to write up to `nbyte` bytes into file descriptor `fildes` from the buffer starting at `buf`.
+
+The return value will be the number of bytes written on success.
+
+## `remove`
+
+This function deletes a file specified by `path`.
+
+The return value will be set to `0` if the operation successes. If it fails, the value will be `-1`.
```
Comment (DavidSpickett) [inline demos/file/README.md]: "if the operation succeeds"

File: demos/file/README.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+## Walkthrough
+
+The `setup` creates 5 threads. Each thread demonstrates a scenario.
+
+The first case is `reader`. The thread reads 100 bytes from the source of this demo. Then it replaces the first `'\n'` with a null character. So the output of this case will be `/* Hello file demo! */`.
+
+The second case tries to open a file that is not existing in the path. As a result, the returned file descriptor must be `-1`, which indicates the failure of open.
```
Comment (DavidSpickett) [inline demos/file/README.md]: "that does not exist."

File: demos/file/README.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+For the third case, we test the `write()` by writing a string into a temporary file and verify the operation by justifying the number of bytes we just wrote. After that, we close the file descriptor with `close()`.
```
Comment (DavidSpickett) [inline demos/file/README.md]: "justifying" isn't the word here.

Perhaps:
"and asserting that `write` returned the number of bytes we tried to write."

File: demos/file/README.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+`read()` attempts to read at most `nbyte` bytes from file descriptor `fildes` into the buffer starting at `buf`.
+
+On success, the number of bytes read is returned, and the file position is advanced by this number.
+
+## `write`
+
+We can use `write()` to write up to `nbyte` bytes into file descriptor `fildes` from the buffer starting at `buf`.
+
+The return value will be the number of bytes written on success.
```
Comment (DavidSpickett) [inline demos/file/README.md]: This leaves out the failure value, which is -1 like the rest. (https://man7.org/linux/man-pages/man2/write.2.html)

**Review by DavidSpickett — CHANGES_REQUESTED** [review]

File: docs/demos.md (hunk 1 above)
Comment (OscarShiang) [reply docs/demos.md]: Sorry about that. It seems my vim automatically add a newline at the end of the file.

**Comment by OscarShiang** [comment]
OK, I have updated the commit.

**Review by DavidSpickett — APPROVED** [review]
LGTM

**Pull request merged**